README.md
# <img src="https://avatars2.githubusercontent.com/u/29560114?s=24&v=4"> The Token Store Public API

[![specification](https://img.shields.io/badge/raml-1.0-blue.svg)](https://github.com/raml-org/raml-spec/blob/master/versions/raml-10/raml-10.md/)
[![Join our telegram group!](https://img.shields.io/badge/telegram-@thetokenstore-32a2da.svg)](https://telegram.me/thetokenstore)


### What is [token.store](https://token.store):

- The Token Store is a decentralized exchange built on smart contracts that lets you trade ERC20 (and some ERC223) tokens against Ethereum.
- You can find more information about our fees [here.](https://token.store/fees)
- Our [solidity contract ](https://github.com/tokenstore/contract) is published at [0x1cE7AE555139c5EF5A57CC8d814a867ee6Ee33D8](https://etherscan.io/address/0x1ce7ae555139c5ef5a57cc8d814a867ee6ee33d8#code).
- You can trade tokens both by symbol and contract address.

### REST API

The Base URL is [https://v1-v.api.token.store](https://v1-1.api.token.store). At this endpoint we have [documentation](https://v1-1.api.token.store) generated from  [api-v1-1.raml](api-v1-1.raml)

We use RAML 1.0 for specification. You can use it to automatically generate documentation/client/mocks/etc. You can read more about RAML at [raml.org](https://raml.org/)

### Example usage

- Example for [node.js](bots-example/js/)

### Tools

- [order signer](https://www.npmjs.com/package/@token.store/ethjs-order-signer)
- [ethereumjs-util](https://www.npmjs.com/package/ethereumjs-util)

## FAQ

#### Your API responds with a 400 code

This error means your call has wrong payload. We advise the following steps:

1. Check payload example in our specification.
2. Make sure you're using the right function from our contract to generate the message for signing.
3. Feel free to open an issue here with some example code so we can help you out -- but be careful with your private keys!

#### Your API responds with a 502 code

This error shows that your transaction doesn't approved by our Ethereum node. We advise the following steps:

1. Check the error message! Usually it has all required information to fix the error.
2. Check your account balance. Do you have enough Ethereum for gas payment?
3. Check your nonce. You can use our endpoint `/accounts/{account}` for recieving the recommended nonce for your account.

#### My transaction takes too long to process. Could you help me?

Nope. Every transaction is processed on Ethereum blockchain. Please, use more gas!

#### Do you have different protocols like WebSockets, JSON-RPC, GraphQL?

Not yet -- but please feel free to open a feature request here!

#### Could I have support?

If you have developer request, then please open an issue here.

If you need help with the product, then our [telegram chanel](https://telegram.me/thetokenstore) is the best place for that. You can also email support@token.store for complex queries.


CHANGELOG.md
# Changelog

All notable changes to The Token Store's API will be documented here.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.0] - 2018-06-14
### Deprecated
* `https://v1.api.token.store`. Please use `https://v1-1.api.token.store`
### Changed
* `https://v1.api.token.store` => `https://v1-1.api.token.store`
* GET `https://v1-1.api.token.store/orderbook?pair=<pair>` => GET `https://v1-1.api.token.store/orderbook/{pair}`. Now, pair can have *ETH* both before and after the token symbol. For example, *ETH_EVN* and *EVN_ETH* are valid pairs.
* GET `https://v1-1.api.token.store/orderbook/{pair}` payload changed
from
```
{
  "bids": [[0.0028555, 22], [0.0028, 4]],
  "asks": [[0.345, 0.345], [0.35, 1.20085]]
}
```
to
```
{
  "bids": [
    { price: 0.0028555, volume: 22 },
    { price: 0.0028, volume: 4 }
  ],
  "asks": [
    { price: 0.345, volume: 0.345 },
    { price: 0.35, volume: 1.20085 }
  ]
}
```
### Fixed
* GET `https://v1-1.api.token.store/api/trades` and GET `https://v1.api.token.store/api/trades/{id}` - has the corrected *type* for trades. Before it was wrong.
* GET `https://v1-1.api.token.store/orderbook/{pair}` - returns *volume* in tokens units for bids and asks. Before it was tokens unit for bids and Ethereum unit for asks.

## [1.0.0] - 2018-05-11
### Deprecated
* `https://token.store/api`
* `https://api.token.store/public`
### Changed
* GET `https://token.store/api/ticker` => GET `https://v1.api.token.store/ticker`
* GET `https://token.store/api/trades` => GET `https://v1.api.token.store/trades`
* GET `https://token.store/api/pairs` => GET `https://v1.api.token.store/pairs`
* GET `https://token.store/api/orderbook?pair=<pair>` => GET `https://v1.api.token.store/orderbook?pair=<pair>`
### Added
* GET `https://v1.api.token.store/ticker` - Return config including fee, minOrderSize, blockNumber
* GET `https://v1.api.token.store/contract` - Return contract abis and address
* GET `https://v1.api.token.store/orders` - Provide all active orders ordered by created time
* POST `https://v1.api.token.store/orders` - Create new order
* GET `https://v1.api.token.store/orders/{id}` - Provide infomation for single order including not an order that is not active
* POST `https://v1.api.token.store/orders/{id}/cancel` - Used for send cancel order transaction. Function *cancelOrder* from our contract should be used generating the signature.
* POST `https://v1.api.token.store/api/trades` - Create trade for the order with number orderId from payload. Function *trade* from our contract should be used generating the signature.
* GET `https://v1.api.token.store/api/trades/{id}` - Provide infomation for single trade
* GET `https://v1.api.token.store/accounts/{account}` - Provide information required for making solidity function request for specific address
* GET `https://v1.api.token.store/accounts/{account}/balances` - Return balance for provided tokens
* POST `https://v1.api.token.store/accounts/{account}/deposit` - Deposit Ethereum or a token to smart contract via API. Function *deposit* or *depositToken* from our contract should be used generating the signature.
* POST `https://v1.api.token.store/accounts/{account}/withdraw` - Withdraw Ethereum or a token to smart contract via API. Function *withdraw* or *withdrawToken* from our contract should be used generating the signature.

## [0.x] - 2018-03-27
### Added
* GET `https://token.store/api/ticker` - Returns 24h trade volume and the last price for each coin.
* GET `https://token.store/api/trades` - Return historical trades executed
* GET `https://token.store/api/pairs` - Get list of pairs
* GET `https://token.store/api/orderbook?pair=<pair>` - Orderbook for a pair


bots-examples/js/src/app.js
const Client = require('./client');

const privateKey = 'd38127980869e585639efc7b8d9a30643b486ba62d265fbf8809de9f4fc42c6b';

const client = new Client(privateKey);

async function start () {
  await client.init();
  await client.sync();
  await client.placeBuyOrder('0xd780ae2bf04cd96e577d3d014762f831d97129d0', 18, 0.0005, 1000);
  await client.cancelOrder(109231);
  await client.depositToken('0xabfef126779e80b2f271f12cdd142f2ace540ccf', 10000 * 10 ** 18);
  await client.depositEth(0.1 * 10 ** 18);
  await client.withdrawEth(0.01 * 10 ** 18);
  await client.withdrawToken('0xabfef126779e80b2f271f12cdd142f2ace540ccf', 1000 * 10 ** 18);
  await client.trade(109231, 0.0001 * 10 ** 18);
}

start()
  .then(() => {
    console.log('Done');
  })
  .catch((error) => {
    console.error(error);
  });


bots-examples/js/src/client.js
const {
  privateToAddress,
  zeroAddress
} = require('ethereumjs-util');

const SolidityFunction = require('web3/lib/web3/function');
const { createSignature } = require('@token.store/ethjs-order-signer');
const { toHex } = require('web3').prototype;
const Tx = require('ethereumjs-tx');
const got = require('got');
const { HTTPError } = require('got/errors');
const BigNumber = require('bignumber.js');
BigNumber.config({ EXPONENTIAL_AT: 100 });

const API = 'https://v1.api.token.store';

class Client {
  constructor (secret) {
    this._dateCreators = {};

    this.private = Buffer.from(secret, 'hex');
    this.address = `0x${privateToAddress(this.private).toString('hex')}`;

    this.contractAddress = undefined;

    this.nonce = undefined;
    this.gasPrice = undefined;
    this.gasLimit = undefined;
    this.blockNumber = undefined;
  }

  /**
   * Set fee info, contractAddress and parse abi
   * @returns {Promise<void>}
   */
  async init () {
    const url = `${API}/contract`;
    const data = await this._call(url);

    this.contractAddress = data.contract.address;

    data.contract.abi
      .filter(abi => abi.type === 'function')
      .forEach(abi => {
        const solidityFunc = new SolidityFunction('', abi, '');
        this._dateCreators[abi.name] = (data) => {
          const params = [];
          for (const input of abi.inputs) {
            const param = data[input.name];
            params.push(param);
          }
          return solidityFunc.toPayload(params).data;
        };
      });
  }

  /**
   * Set nonce and gas data. Recommended to run before transaction
   * @returns {Promise<void>}
   */
  async sync () {
    const url = `${API}/config`;
    const data = await this._call(url);

    this.gasPrice = data.recommendedGasPrice;
    this.gasLimit = data.recommendedGasLimit;
    this.blockNumber = data.blockNumber;

    const personalUrl = `${API}/accounts/${this.address}`;
    const { nonce } = await this._call(personalUrl);
    this.nonce = nonce;
  }

  /**
   * Withdraw etherium.
   * @param {string} amount - etherium in wei.
   * @returns {Promise<*>}
   */
  async withdrawEth (amount) {
    const url = `${API}/accounts/${this.address}/withdraw`;
    const data = this._dateCreators['withdraw']({
      _amount: amount
    });

    const payload = this._createBasePayload(data);
    payload.tokenAddress = zeroAddress();
    payload.amount = amount;
    return this._call(url, payload);
  }

  /**
   * Withdraw token
   * @param {string} _token - token address
   * @param {string} _amount - in uint256
   * @returns {Promise<*>}
   */
  async withdrawToken (_token, _amount) {
    const url = `${API}/accounts/${this.address}/withdraw`;
    const data = this._dateCreators['withdrawToken']({
      _amount,
      _token
    });

    const payload = this._createBasePayload(data);
    payload.tokenAddress = _token;
    payload.amount = _amount;
    return this._call(url, payload);
  }

  /**
   * Deposit etherium
   * @param {string} amount - etherium in wei.
   * @returns {Promise<*>}
   */
  async depositEth (amount) {
    const url = `${API}/accounts/${this.address}/deposit`;
    const data = this._dateCreators['deposit']({});

    const payload = this._createBasePayload(data, amount);
    payload.tokenAddress = zeroAddress();
    payload.amount = amount;
    return this._call(url, payload);
  }

  /**
   * Deposit token
   * @param {string} _token - token address
   * @param {string} amount - in uint256
   * @returns {Promise<*>}
   */
  async depositToken (_token, _amount) {
    const url = `${API}/accounts/${this.address}/deposit`;
    const data = this._dateCreators['depositToken']({
      _token,
      _amount
    });

    const payload = this._createBasePayload(data);
    payload.tokenAddress = _token;
    payload.amount = _amount;
    return this._call(url, payload);
  }

  /**
   * @param {string} tokenAddress
   * @param {number} tokenDecimals
   * @param {number} price - etherium per token
   * @param {number} amount - token amount
   * @param {number} expiresIn - after what block the order will not be valid anymore
   * @returns {Promise<OrderObject>}
   */
  async placeBuyOrder (tokenAddress, tokenDecimals, price, amount, expiresIn = 10000) {
    const url = `${API}/orders`;
    const payload = this._createOrder('buy', tokenAddress, tokenDecimals, price, amount, this.blockNumber + expiresIn);

    return this._call(url, payload, 'POST');
  }

  /**
   * @param {string} tokenAddress
   * @param {number} tokenDecimals
   * @param {number} price - etherium per token
   * @param {number} amount - token amount
   * @param {number} expiresIn - after what block the order will not be valid anymore
   * @returns {Promise<OrderObject>}
   */
  async placeSellOrder (tokenAddress, tokenDecimals, price, amount, expiresIn = 10000) {
    const url = `${API}/orders`;
    const payload = this._createOrder('sell', tokenAddress, tokenDecimals, price, amount, this.blockNumber + expiresIn);

    return this._call(url, payload, 'POST');
  }

  /**
   * @param {number} orderId - in /orders/ endpoint
   * @returns {Promise<void>}
   */
  async cancelOrder (orderId) {
    const orderUrl = `${API}/orders/${orderId}`;
    const order = await this._call(orderUrl);

    const data = this._dateCreators['cancelOrder']({
      _tokenGet: order.tokenGet,
      _amountGet: order.amountGet,
      _tokenGive: order.tokenGive,
      _amountGive: order.amountGive,
      _expires: order.expires,
      _nonce: order.nonce,
      _v: order.signature.v,
      _r: order.signature.r,
      _s: order.signature.s
    });

    const payload = this._createBasePayload(data);
    return this._call(`${orderUrl}/cancel`, payload);
  }

  /**
   * @param {number} orderId - in /orders/ endpoint
   * @param {number} amount - token amount
   * @returns {Promise<void>}
   */
  async trade (orderId, amount) {
    const tradesUrl = `${API}/trades`;
    const orderUrl = `${API}/orders/${orderId}`;
    const order = await this._call(orderUrl);

    const data = this._dateCreators['trade']({
      _tokenGet: order.tokenGet,
      _amountGet: order.amountGet,
      _tokenGive: order.tokenGive,
      _amountGive: order.amountGive,
      _expires: order.expires,
      _nonce: order.nonce,
      _v: order.signature.v,
      _r: order.signature.r,
      _s: order.signature.s,
      _amount: amount
    });

    const payload = this._createBasePayload(data);
    payload.orderId = orderId;
    payload.account = this.address;
    payload.amount = amount;

    return this._call(tradesUrl, payload);
  }

  /**
   * @param {string} url
   * @param body
   * @private
   */
  async _call (url, body = undefined) {
    try {
      const response = await got(url, {
        json: true,
        method: body ? 'POST' : 'GET',
        body,
        headers: {
          'user-agent': 'ts-bot v1.0.0'
        }
      });
      return response.body;
    } catch (error) {
      if (error instanceof HTTPError) {
        console.log(error);
        throw new Error(error.response.body.message);
      } else {
        throw error;
      }
    }
  }

  /**
   * @param {string} data
   * @param {number|string} value
   * @returns {{tx: {nonce: number, gasPrice: number, gasLimit: number}, signature: {v: number, r: string, s: string}}}
   * @private
   */
  _createBasePayload (data, value = 0) {
    const tx = new Tx({
      nonce: this.nonce,
      gasPrice: this.gasPrice,
      gasLimit: this.gasLimit,
      to: this.contractAddress,
      value: value && toHex(value),
      data
    });

    tx.sign(this.private);

    return {
      tx: {
        nonce: this.nonce,
        gasPrice: this.gasPrice,
        gasLimit: this.gasLimit
      },
      signature: {
        r: tx.r.toString('hex'),
        s: tx.s.toString('hex'),
        v: parseInt(tx.v.toString('hex'), 16)
      }
    };
  }

  _toWei (eth, decimals) {
    return (new BigNumber(String(eth))).times(new BigNumber(10 ** decimals)).floor();
  }

  /**
   * @typedef {Object} OrderObjectSignature
   * @property {number} v
   * @property {string} r
   * @property {string} s
   */

  /**
   * @typedef {Object} OrderObject
   * @property {string} account
   * @property {string} contract
   * @property {string} tokenGet
   * @property {string} amountGet
   * @property {string} tokenGive
   * @property {string} amountGive
   * @property {number} nonce
   * @property {number} expires
   * @property {OrderObjectSignature} signature
   */

  /**
   * @private
   * @param {string} side - buy/sell
   * @param {string} tokenAddress
   * @param {string} tokenDecimals
   * @param {number} price
   * @param {number} amount
   * @param {number} expires
   * @returns OrderObject
   * @see https://github.com/etherdelta/bots/blob/master/js/service.js
   */
  _createOrder (side, tokenAddress, tokenDecimals, price, amount, expires) {
    if (side !== 'buy' && side !== 'sell') throw new Error('Side must be buy or sell');

    const amountBigNum = new BigNumber(String(amount));
    const amountBaseBigNum = new BigNumber(String(amount * price));
    const amountGet = side === 'buy'
      ? this._toWei(amountBigNum, tokenDecimals)
      : this._toWei(amountBaseBigNum, 18);
    const amountGive = side === 'sell'
      ? this._toWei(amountBigNum, tokenDecimals)
      : this._toWei(amountBaseBigNum, 18);

    const orderNonce = Number(Math.random().toString().slice(2));

    const order = {
      account: this.address,
      contract: this.contractAddress,
      tokenGet: side === 'buy' ? tokenAddress : zeroAddress(),
      amountGet: amountGet.toString(),
      tokenGive: side === 'sell' ? tokenAddress : zeroAddress(),
      amountGive: amountGive.toString(),
      nonce: orderNonce,
      expires
    };

    order.signature = createSignature(order, this.private);

    return order;
  }
}

module.exports = Client;
